File: PD Project/Firmware/PD_VRMT_v1_0_6_20240814/main/system.cpp
#include <Arduino.h>
#include "system.h"
#include <avr/wdt.h>

System::System() {};

void System::init(int pin_batt, int pin_charge, int pin_charge_led, int pin_power_supply) {
    _pin_batt = pin_batt;
    _pin_charge = pin_charge;
    _pin_charge_led = pin_charge_led;
    _pin_power_supply = pin_power_supply;
    pinMode(_pin_charge_led, OUTPUT);
    pinMode(_pin_power_supply, OUTPUT);

    digitalWrite(_pin_power_supply, HIGH);
}

void System::check_battery() {
    _voltage = analogRead(_pin_batt) * BATTERY_ADC;

    _charging = analogRead(_pin_charge);
    bool currently_is_charging = (_charging >= 200);
    if (currently_is_charging != _is_charging) {
        digitalWrite(_pin_charge_led, (currently_is_charging) ? HIGH : LOW);
        _is_charging = currently_is_charging;
    }
}

bool System::check_time(unsigned long* timer, unsigned long period) {
    // update the time passed since the last timer check
    // notify if a particular time period has passed
    _dt = micros() - *timer;
    if (_dt >= period) {
        // update the timer at the timed interval
        *timer = micros();
        return true;
    }
    return false;
}

float System::get_voltage() {
    return _voltage;
}

float System::get_charging() {
    return _charging;
}

File: PD Project/python_code/system.py
import time
import RPi.GPIO as GPIO
from smbus import SMBus# Need ADC or other method to read analog data

class System:
    # Battery ADC conversion factor 
    BATTERY_ADC = 0.01807
    
    def __init__(self):
        self._pin_batt = None
        self._pin_charge = None
        self._pin_charge_led = None
        self._pin_power_supply = None
        self._is_charging = False
        self._dt = 0.0
        self._voltage = 0.0
        self._charging = 0.0
        self._i2c_bus = SMBus(1)  # For I2C communication (if needed)# Need ADC or other method to read analog data
        
    def init(self, pin_batt, pin_charge, pin_charge_led, pin_power_supply):
        """Initialize GPIO pins and system monitoring"""
        self._pin_batt = pin_batt
        self._pin_charge = pin_charge
        self._pin_charge_led = pin_charge_led
        self._pin_power_supply = pin_power_supply
        
        GPIO.setmode(GPIO.BCM)
        GPIO.setup(self._pin_charge_led, GPIO.OUT)
        GPIO.setup(self._pin_power_supply, GPIO.OUT)
        
        # Enable power supply by default
        GPIO.output(self._pin_power_supply, GPIO.HIGH)
        
        # Setup ADC for battery monitoring (using MCP3008 or similar)
        # You'll need to implement your ADC reading method
        
    def check_battery(self):
        """Check battery voltage and charging status"""
        try:
            # Read battery voltage (implement your ADC reading method)
            raw_value = self._read_adc(self._pin_batt)
            self._voltage = raw_value * self.BATTERY_ADC
            
            # Read charging status
            raw_charge = self._read_adc(self._pin_charge)
            self._charging = raw_charge
            currently_is_charging = (raw_charge >= 200)
            
            if currently_is_charging != self._is_charging:
                GPIO.output(self._pin_charge_led, 
                          GPIO.HIGH if currently_is_charging else GPIO.LOW)
                self._is_charging = currently_is_charging
                
        except Exception as e:
            print(f"Battery check error: {e}")

    def _read_adc(self, channel):# Need ADC or other method to read analog data
        """Read from ADC (MCP3008 implementation example)"""
        # This is a placeholder - implement based on your ADC hardware
        # Example for MCP3008:
        cmd = 0b11 << 6  # Start bit + single ended
        cmd |= (channel & 0x07) << 3  # Channel number
        data = self._i2c_bus.write_i2c_block_data(0x08, cmd, [0])
        return ((data[0] & 0x03) << 8) | data[1]  # Combine bytes
        
    def check_time(self, timer, period):
        """Check if specified time period has passed (in microseconds)"""
        current_time = time.time() * 1000000  # Convert to microseconds
        self._dt = current_time - timer[0]
        
        if self._dt >= period:
            timer[0] = current_time  # Update the timer
            return True
        return False

    def get_voltage(self):
        """Get current battery voltage"""
        return self._voltage

    def get_charging(self):
        """Get current charging status value"""
        return self._charging

    def cleanup(self):
        """Clean up GPIO resources"""
        GPIO.cleanup()